subsys/lorawan/services/multicast.c

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright (c) 2022 Martin Jäger <martin@libre.solar>
+ * Copyright (c) 2022 tado GmbH
+ *
+ * Parts of this implementation were inspired by LmhpRemoteMcastSetup.c from the
+ * LoRaMac-node firmware repository https://github.com/Lora-net/LoRaMac-node
+ * written by Miguel Luis (Semtech).
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "lorawan_services.h"
+#include "../lw_priv.h"
+
+#include <LoRaMac.h>
+#include <zephyr/lorawan/lorawan.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/random/random.h>
+#include <zephyr/settings/settings.h>
```

Comment:
Persistent state management should be done through the `lorawan/lorawan_nvm.h` header to enable people who don't use the settings API to manage it through some other means.

Reply:
I realized that the existing settings-based NVM backend already stores the multicast session details (hidden in `LORAMAC_NVM_NOTIFY_FLAG_MAC_GROUP2`), so I removed all the custom session data storage code.